File: Data Structures & Algorithms/generate-parentheses/submission-0.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []

        def dfs(curr, o, c):
            if o == 0 and c == 0:
                result.append(curr)
                return
            if o == 0:
                for _ in range(c):
                    curr += ")"
                result.append(curr)
                return
            if o == c:
                curr += "("
                dfs(curr, o-1, c)
            else:
                curr += "("
                dfs(curr, o-1, c)
                curr = curr[:-1]
                curr += ")"
                dfs(curr, o, c-1)
        
        dfs("", n, n)
        return result

File: Data Structures & Algorithms/clone-graph/submission-0.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (node == nullptr) return nullptr;

        unordered_map<Node*, Node*> clonedMap;

        function<Node*(Node*)> dfs = [&](Node* curr) -> Node* {
            
            if (clonedMap.count(curr)) {
                return clonedMap[curr];
            }

            Node* copy = new Node(curr->val);

            clonedMap[curr] = copy;
            for (Node* neighbor : curr->neighbors) {
                copy->neighbors.push_back(dfs(neighbor));
            }

            return copy;
        };

        return dfs(node);
    }
};
